### PR TITLE
CLI: Fix `upgrade` detecting the wrong version of existing Storybooks

### LINKED ISSUE #25734: [Bug]: `upgrade` command always throws `UpgradeStorybookToSameVersionError` when using `pnpx`
### Describe the bug

Running pnpm dlx storybook@next upgrade in @storybook/addon-onboarding throws `UpgradeStorybookToSameVersionError`, although the project uses Storybook dependencies with the version 7.1.0-alpha.38. Everything works fine when using `npx storybook@next` upgrade.

### To Reproduce

1. `npx storybook@latest sandbox`
2. cd into sandbox
3. `pnpm install`
4. `pnpx storybook@next upgrade`
5. Get error

### System

_No response_

### Additional context

_Originally posted by @valentinpalkovic in https://github.com/storybookjs/storybook/discussions/25692#discussioncomment-8230065_

This happens because we use `require.resolve('@storybook/cli/package.json')` to get the actual package path, and then check the `version` field in the `package.json`. However `require.resolve` in `pnpx` will get its own package and not the one in the actual working directory.

We need to change the logic to the package manager's CLI what version is installed in CWD (eg. `pnpm why @storybook/cli`), and then parse that output to determine the current version.

### PR DESCRIPTION
**Pull request by JReinhold**
Closes #25734 

## What I did

Changed the `upgrade` command to use package manager's CLI for detecting the existing version of _both_ `storybook` and `@storybook/cli` instead of using a combination of `require.resolve` and reading `package.json`, as that would run in the global context thus not detecting the correct version as it is _in the project_. The `packageManager.findInstallations` function being used was originally develop for the `doctor` command but it works fine in this context too.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

This can be (and has been) tested with the canary release and using the different package manager executors. Eg. in a 7.6.10 sandbox/project, running this should result in the message _Upgrading Storybook from version **7.6.10** to version **0.0.0-pr-25752-sha-424630b7**.._

- `npx      storybook@0.0.0-pr-25752-sha-424630b7 upgrade`
- `yarn dlx storybook@0.0.0-pr-25752-sha-424630b7 upgrade`
- `pnpm dlx storybook@0.0.0-pr-25752-sha-424630b7 upgrade`
- `pnpx     storybook@0.0.0-pr-25752-sha-424630b7 upgrade`

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-25752-sha-424630b7`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-25752-sha-424630b7 sandbox` or in an existing project with `npx storybook@0.0.0-pr-25752-sha-424630b7 upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-25752-sha-424630b7`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-25752-sha-424630b7) |
| **Triggered by** | @JReinhold |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`jeppe/fix-upgrade-version-detection`](https://github.com/storybookjs/storybook/tree/jeppe/fix-upgrade-version-detection) |
| **Commit** | [`424630b7`](https://github.com/storybookjs/storybook/commit/424630b7c6ad65e2ff293c852f7004b2e9a07cc7) |
| **Datetime** | Fri Jan 26 09:33:43 UTC 2024 (`1706261623`) |
| **Workflow run** | [7666315388](https://github.com/storybookjs/storybook/actions/runs/7666315388) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=25752`_
</details>

### CODE REVIEW

File: code/lib/telemetry/src/package-json.ts [hunk 1]
```diff
@@ -27,9 +27,3 @@ export const getActualPackageJson = async (packageName: string) => {
   const packageJson = await fs.readJson(resolvedPackageJson);
   return packageJson;
 };
-
-// Note that this probably doesn't work in Yarn PNP mode because @storybook/telemetry doesn't depend on @storybook/cli
-export const getStorybookCoreVersion = async () => {
-  const { version } = await getActualPackageVersion('@storybook/cli');
-  return version;
-};
```
Comment (JReinhold) [inline code/lib/telemetry/src/package-json.ts:35]: @shilman this is a function in telemetry but it was only used by `upgrade` which doesn't use it anymore, so I felt it was okay to remove it. Let me know if not.

**Review by ndelangen — APPROVED** [review]

File: code/lib/cli/src/upgrade.ts [hunk 2]
```diff
@@ -37,6 +37,18 @@ export const getStorybookVersion = (line: string) => {
   };
 };
 
+const getInstalledStorybookVersion = async (packageManager: JsPackageManager) => {
+  const installations = await packageManager.findInstallations(['storybook', '@storybook/cli']);
+  if (!installations) {
+    return;
+  }
+  const cliVersion = installations.dependencies['@storybook/cli']?.[0].version;
```
Comment (valentinpalkovic) [inline code/lib/cli/src/upgrade.ts:45]: `installations.dependencies['@storybook/cli']?.[0]` -> Can this be undefined? If so `installations.dependencies['@storybook/cli']?.[0].version` will through.

Comment (JReinhold) [reply code/lib/cli/src/upgrade.ts:45]: It can't, TS would complain, and the current code producing it doesn't do that either.

**Review by yannbf — APPROVED** [review]
I'm assuming you tested the scenarios thoroughly via the canary, in that case LGTM!

**Pull request merged**